packages/core/mx/contracts/escrow/src/lib.rs

```diff
@@ -0,0 +1,354 @@
+            aggregate_bulk_amount += amount;
+        }
+
+        require!(aggregate_bulk_amount < self.bulk_max_value().get(), "Bulk value too high");
+        if balance < aggregate_bulk_amount {
+            return bulk_paid
+        }
+
+        self.store_results(&url, &hash);
+
+        let (reputation_oracle_fee, recording_oracle_fee, final_amounts) = self.finalize_payouts(&amounts);
+
+        for (i, recipient) in recipients.clone().into_iter().enumerate() {
+            let amount = final_amounts.get(i);
+            if *amount == 0 {
+                continue
+            }
+            self.send().direct_esdt(&recipient, &self.token().get(), 0, &amount);
+        }
+
+        self.send().direct_esdt(&self.reputation_oracle().get(), &self.token().get(), 0, &reputation_oracle_fee);
+        self.send().direct_esdt(&self.recording_oracle().get(), &self.token().get(), 0, &recording_oracle_fee);
+
+        bulk_paid = true;
+        balance = self.get_balance();
+        let mut status = self.get_status();
+        if bulk_paid {
+            if status == EscrowStatus::Pending {
+                self.remaining_solutions().update(|x| *x -= recipients.len() as u64);
+                status = EscrowStatus::Partial;
+            }
+
+            if balance > 0 && status == EscrowStatus::Partial && self.remaining_solutions().get() == 0 {
+                self.send().direct_esdt(&self.canceler().get(), &self.token().get(), 0, &balance);
+                status = EscrowStatus::Paid;
+            }
+
+            if balance == 0 && status == EscrowStatus::Partial {
+                status = EscrowStatus::Paid;
+            }
+
+            self.status().set(status);
+        }
+
+        self.bulk_transfer_event(tx_id, recipients.len() as u64);
+        true
+    }
+
+    fn finalize_payouts(&self, amounts: &MultiValueEncoded<BigUint>) -> (BigUint, BigUint, ManagedVec<BigUint>){
+        let mut reputation_oracle_fee = BigUint::zero();
+        let mut recording_oracle_fee = BigUint::zero();
+
+        let mut final_amounts: ManagedVec<BigUint> = ManagedVec::new();
+
+        for given_amount in amounts.clone().into_iter() {
+            let single_reputation_oracle_fee = self.reputation_oracle_stake().get() * &given_amount / BigUint::from(100_u64);
+            let single_recording_oracle_fee = self.recording_oracle_stake().get() * &given_amount / BigUint::from(100_u64);
+            let amount = given_amount - &single_reputation_oracle_fee - &single_recording_oracle_fee;
+            reputation_oracle_fee = &reputation_oracle_fee + &single_reputation_oracle_fee;
+            recording_oracle_fee = &recording_oracle_fee + &single_recording_oracle_fee;
+
+            final_amounts.push(amount);
+        }
+
+        (reputation_oracle_fee, recording_oracle_fee, final_amounts)
+    }
+
+    fn store_results(&self, url: &ManagedBuffer, hash: &ManagedBuffer) {
+        let write_on_chain = url.len() != 0 || hash.len() != 0;
```

Comment:
to have parity with the sol contracts, I intentionally left the same condition there. If the team decides this can be fixed, both contracts should be updated